ecl/eclscheduler/eclscheduler.cpp

```diff
@@ -180,6 +181,258 @@ version: "1.0"
+
+        while (running)
+        {
+            // wait for either a config change or an abort
+            waiter.wait();
+
+            if (running)
+            {
+                // Process a config change if one is pending
+                try
+                {
+                    CriticalBlock b(updateSchedulersCS);
+                    // onAbort could have triggered before or during the above switch, if so, we do no want to connect/block on new queue
+                    if (!running)
+                        break;
+
+                    bool newQueues = false;
+                    if (updatedQueueNames)
+                    {
+                        currentQueueNames.set(updatedQueueNames);
+                        updatedQueueNames.clear();
+                        newQueues = true;
+                    }
+
+                    if (newQueues)
+                        updateSchedulerQueues();
+                }
+                catch (IException *E)
+                {
+                    EXCLOG(E);
+                    releaseAtoms();
+                    ExitModuleObjects();
+                    _exit(2);
+                }
+                catch (...)
+                {
+                    IERRLOG("Unknown exception caught in eclscheduler processQueues - restarting");
+                    releaseAtoms();
+                    ExitModuleObjects();
+                    _exit(2);
+                }
+            }
+        }
+        DBGLOG("eclscheduler closing");
+        stop();
+    }
+
+private:
+    StringBuffer &getQueues(const IPropertyTree *config, StringBuffer &queueNameList)
+    {
+#ifdef _CONTAINERIZED
+        Owned<IPTreeIterator> queues = config->getElements("queues");
+        ForEach(*queues)
+        {
+            IPTree &queue = queues->query();
+            const char *qname = queue.queryProp("@name");
+            if (queueNameList.length())
+                queueNameList.append(",");
+            queueNameList.append(qname);
+        }
+#else
+        const char *processName = config->queryProp("@name");
+        queueNameList.append(processName);
+#endif
+        return queueNameList;
+    }
+
+    void updateSchedulerQueues()
+    {
+        std::unordered_set<std::string> updatedQueuesSet = splitBy(currentQueueNames.str(), ",");
+
+        // stop and remove schedulers not in the new list as they are no longer needed
+        // for those in the new list, if they are already running, leave them alone
+        std::list<aindex_t> schedulersToBeRemoved;
+        ForEachItemIn(schedIdx, schedulers)
+        {
+            if (updatedQueuesSet.find(schedulers.item(schedIdx).getServerName().str()) == updatedQueuesSet.end())
+            {
+                // Queue has been removed
+                DBGLOG("Stopped listening to queue %s", schedulers.item(schedIdx).getServerName().str());
+                schedulers.item(schedIdx).stop();
+                schedulersToBeRemoved.emplace_front(schedIdx);
```

Comment:
could remove from 'schedulers' array directly (with schedulers.pop()), if you traverse the array in reverse.. with ForEachItemInRev

Reply:
not pop(), because won't always be last element - use remove(schedIdx);